book/src/implemented-proposals/abi-management.md

```diff
@@ -73,16 +73,84 @@ digest from the assertion test error message.
 
 In general, once we add `frozen_abi` and its change is published in the stable
 release channel, its digest should never change. If such a change is needed, we
-should opt for defining a new struct like `FooV1`. And special release flow like
-hard forks should be approached.
+should opt for defining a new `struct` like `FooV1`. And special release flow
+like hard forks should be approached.
 
 # Implementation remarks
 
 We use some degree of macro machinery to automatically generate unit tests
 and calculate a digest from ABI items. This is doable by clever use of
-`serde::Serialize` ([1]) and `any::typename` ([2]). For a precedent for similar
+`serde::Serialize` ([1]) and `any::type_name` ([2]). For a precedent for similar
 implementation, `ink` from the Parity Technologies [3] could be informational.
 
+# Implementation details
+
+The implementation's goal is to detect unintended ABI changes automatically as
+much as possible. To that end, the digest of structural ABI information is
+calculated with best-effort accuracy and stability.
+
+When the ABI digest check is run, it dynamically computes an ABI digest by
+recursively digesting the ABI of fields of the ABI item, by re-using the
+`serde`'s serialization functionality, proc macro and generic specialization.
+And then, the check `assert!`s that it finalized digest value is identical as
+what is specified in the `frozen_abi` attribute.
+
+To realize that, it creates a sample instance of the type and a custom
+`Serializer` instance for `serde` to recursively traverse its fields as if
+serializing the sample for real. This traversing must be done via `serde` to
+really capture what kinds of data actually would be serialized by `serde`, even
+considering custom non-`derive`d `Serialize` trait implementations.
+
+# The ABI digesting process
+
+This part is a bit complex. There is three inter-depending parts: `AbiSample`,
```

Comment:
The name `AbiSample` is a little rough on my brain.  It's not intuitively obvious to me what types need AbiSample, AbiDigest, or in come cases both.  I suspect this is just a naming thing.

AbiSample: this for any type or dependent type that a part of a frozen abi
AbiDigest: this is the top-level type that's frozen?

Reply:
@mvines I've gone though much thought and I've finally came up with better naming:

`AbiExample`: this for any type or dependent type that's a part of a frozen abi
`AbiEnumVisitor`: I've clarified the implementation and make this only applicable to `enum`s.